functional_programming_tests/test/without_functional_programming.cpp

```diff
@@ -0,0 +1,205 @@
+#include <gtest/gtest.h>
+
+#include <cmath>
+#include <iostream>
+#include <memory>
+#include <optional>
+#include <vector>
+
+#include <rclcpp/rclcpp.hpp>
+#include <std_msgs/msg/u_int8_multi_array.hpp>
+
+using namespace std::chrono_literals;
+
+// production_code.h/cc
+template<typename T>
+using Map = std::vector<std::vector<T>>;
+
+// X, Y position
+struct Pos {
+	size_t x;
+	size_t y;
+};
+
+using Path = std::vector<Pos>;
+
+// Operator overload to print path
+std::ostream& operator<<(std::ostream& os, std::vector<Pos> const& path) {
+	for (auto const& pose:path) {
+		os << "(" << pose.x << ", " << pose.y << ")\n";
+	}
+	return os;
+}
+
+// Operator overload for position comparison
+bool operator==(Pos const& lhs, Pos const& rhs) {
+	return lhs.x == rhs.x && lhs.y == rhs.y;
+}
+
+class PathGenerator {
+public:
+	PathGenerator() : node_(std::make_shared<rclcpp::Node>("bad_test_subscriber")), executor_(std::make_shared<rclcpp::executors::SingleThreadedExecutor>()) {
```

Comment:
```suggestion
	PathGenerator() : node_{std::make_shared<rclcpp::Node>("bad_test_subscriber")}, executor_{std::make_shared<rclcpp::executors::SingleThreadedExecutor>()} {
```
I think we should use uniform initialization syntax but I won't die on that hill.
Also formatting.
Also, do you have to add an executor in the class?
https://docs.ros.org/en/foxy/Tutorials/Beginner-Client-Libraries/Writing-A-Simple-Cpp-Publisher-And-Subscriber.html
If so, please document why.
Also "bad_test_subscriber" is too judgmental

Reply:
I added the executor so that it would process the callbacks in the background essentially. I think I can work something out where I do a bunch of `spin_once` calls? Maybe a `spin_until_future_complete`. I just need to ensure that the costmap is received by subscriber. Sadly, `spin` just block indefinitely 😢